implementation/src/main/java/io/smallrye/config/KeyMapBackedConfigSource.java

```diff
@@ -0,0 +1,33 @@
+package io.smallrye.config;
+
+import java.util.Map;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+import io.smallrye.config.common.AbstractConfigSource;
+
+public abstract class KeyMapBackedConfigSource extends AbstractConfigSource {
+    private static final long serialVersionUID = 4378754290346888762L;
+
+    private final KeyMap<String> properties;
+
+    public KeyMapBackedConfigSource(final String name, final KeyMap<String> properties) {
+        super(name, ConfigSource.DEFAULT_ORDINAL);
+        this.properties = properties;
+    }
+
+    public KeyMapBackedConfigSource(final String name, final int ordinal, final KeyMap<String> properties) {
+        super(name, ordinal);
+        this.properties = properties;
+    }
+
+    @Override
+    public Map<String, String> getProperties() {
+        return properties.toMap();
```

Comment:
This is not correct for default values.  The way I originally implemented it was correct: backed by `KeyMap`, but `getProperties` (and thus `getPropertyNames`) should be empty.

Reply:
Hum, why do you think that `getProperties` shouldn't also return the property if a default value is set?

Reply:
Multiple reasons:

* Because the property isn't really present (that's why the default value would be returned).
* Because the value returned in this case doesn't actually match the property names that it corresponds to (for example if I have a default value for `foo.*.bar`, then the default value config source does not contain `foo.baz.bar` even though it would return a value for it; OTOH if I have a default value for `foo.bar`, the default value config source *does* contain that property - inconsistency is a form of incorrectness)
* Because it will cause confusion when determining which values are actually present in the configuration, which Quarkus presently relies upon.